client/src/legacy/SudoModePasswordField/SudoModePasswordFieldEntwine.js

```diff
@@ -0,0 +1,63 @@
+/* global window */
+import jQuery from 'jquery';
+import React from 'react';
+import { createRoot } from 'react-dom/client';
+import { loadComponent } from 'lib/Injector';
+import qs from 'qs';
+
+jQuery.entwine('ss', ($) => {
+  $('.js-injector-boot .SudoModePasswordField').entwine({
+    Component: null,
+    ReactRoot: null,
+
+    onmatch() {
+      // onmatch will match both the field holder and the field
+      // we only want to run this on the field holder
+      if (this.is('input')) {
+        return;
+      }
+      this._super();
+      const cmsContent = this.closest('.cms-content').attr('id');
+      const context = (cmsContent)
+        ? { context: cmsContent }
+        : {};
+      const SudoModePasswordField = loadComponent('SudoModePasswordField', context);
+      this.setComponent(SudoModePasswordField);
+      this.refresh();
+    },
+
+    onunmatch() {
+      this._super();
+      const root = this.getReactRoot();
+      if (root) {
+        root.unmount();
+        this.setReactRoot(null);
+      }
+    },
+
+    /**
+     * This is used by the react component to reload the page after sudo mode is activated
+     */
+    reloadPage() {
+      // Add a ?reload=1 query parameter to the URL to force the browser to reload the page
+      // window.location.reload() does not work as expected
+      const query = qs.parse(window.location.search, { ignoreQueryPrefix: true });
+      const reload = query.reload ? parseInt(query.reload, 10) + 1 : 1;
+      const hrefNoQuery = window.location.href.split('?')[0];
+      window.location.href = hrefNoQuery + qs.stringify({ ...query, reload }, { addQueryPrefix: true });
+    },
```

Comment:
This is probably not the right way to handle this - we only need to reload the form/gridfield, not the whole page. We should use one of the `refresh()`, `redraw()`, or `reloadCurrentPanel()` entwine functions - especially when it's for specifically just a gridfield on its own.

For example on a GridField use `refresh()`.
For whole forms, form submission uses `redraw()` which is a shortcut for `refresh()` with a something-extra that it does - I dunno if that something extra is necessary or not - or when `reloadCurrentPanel()` should be used instead. I'd try one of the first two first, and `reloadCurrentPanel()` as a last resort.

Reply:
It would always be reload the entire form, because the sudo password fields are added to the form, i.e. when a relation gridfield is sudo protected you still need to click on a record to get an edit form with a sudo password field

What supposed to be the syntax for doing this? I tried `this.closest('.cms-edit-form').reloadCurrentPanel();` and the same for `refresh` and `redraw` but none of this actually did anything

Reply:
> It would always be reload the entire form, because the sudo password fields are added to the form

Fair enough. However, reloading the form is not the same as reloading the _whole page_ which the current code is doing.

I'll have a quick play and see if I can get a working example of using those functions.

Reply:
Looks like this is the way:
```js
this.closest('.cms-container').reloadCurrentPanel();
```

There is a bug in the `reloadCurrentPanel()` implementation though, so this needs to be done in https://github.com/silverstripe/silverstripe-admin/blob/9bc9af892de849ecdcd1c0a8fbb58d7ce49ed4df/client/src/legacy/LeftAndMain.js#L438-L440 too:

```diff
 reloadCurrentPanel: function() {
-  this.loadPanel(document.URL, null, null, true);
+  this.loadPanel(document.URL, '', {}, true);
 },
```

Reply:
It's nicer this way, though doesn't quite work in all scenarios. When activating sudo mode from a record being edited from a gridfield e.g. a regular member record in security admin, on reload the form is still in read-only mode

Reply:
For me even the current reload page doesn't work for that same scenario.

The reason is that when you click on a record in the gridfield, because the gridfield is set to be "readonly" it uses the `GridFieldViewButton` instead of the `GridFieldEditButton` - so the URL looks like `https://sudowudo.ddev.site/admin/security/users/EditForm/field/users/item/93/view?gridState-users-0=%7B%22Readonly%22%3Atrue%7D` instead of like `https://sudowudo.ddev.site/admin/security/users/EditForm/field/users/item/93/?gridState-users-0=%7B%22Readonly%22%3Atrue%7D` (note the explicit `/view` suffix)

One way to fix this would be to keep the edit button (so we get the correct URL) and trust `GridFieldDetailForm_ItemRequest` to ensure the form itself is in readonly mode.

Reply:
Have updated it so that `/view` is not suffixed, and we can now use `reloadCurrentPanel();`